### PR TITLE
Enqueue transactions off of the consensus thread

### PR DESCRIPTION
**Pull request by mystenmark**
Should greatly reduce consensus CPU utilization.

Question for reviewers: Do we feel like we need to limit the number of pending enqueue task?

### CODE REVIEW

**Review by mwtian — APPROVED** [review]

**Comment by mwtian** [comment]
It is hard for me to see transaction manager enqueue tasks having more backlogs than the pending and executing transactions, which we already use to flow control. Also it does not seem stable to back pressure up to Narwhal consensus output.

**Review by andll — APPROVED** [review]
>Question for reviewers: Do we feel like we need to limit the number of pending enqueue task?

I think if it is easy to do(for ex add semaphore to consensus handler and pass permit to spawned task), then yes, it would be better to cap this

**Comment by andll** [comment]
>Also it does not seem stable to back pressure up to Narwhal consensus output.

We definitely need to be able to provide backpressure to narwhal to handle large spikes, so I am a bit hesitant to remove backpressure just because we think we can't handle it well.

**Comment by mwtian** [comment]
> > Also it does not seem stable to back pressure up to Narwhal consensus output.
> 
> We definitely need to be able to provide backpressure to narwhal to handle large spikes, so I am a bit hesitant to remove backpressure just because we think we can't handle it well.

Backpressure should be done at the "front", e.g. transaction signing, submissions to Sui & to Narwhal. After a transaction has gone through Narwhal, it must be executed and we just have to buffer it somewhere. The enqueue task and txn mgr itself are good places to park spikes of transactions. I'm not sure keeping data inside Narwhal is better. Later we can keep a count of enqueue tasks and use that to signal back pressure for signing and submission too.

**Comment by mwtian** [comment]
We can increment an atomic counter of enqueuing transactions and include its value in the sum when calculating `execution_queue_len()`. This will provide back pressure at signing and submission time.

**Comment by mystenmark** [comment]
I think I'm most convinced by @mwtian's argument: TransactionManager is really just a buffer.

**Comment by andll** [comment]
I still think all parts of the system need to support back pressure. We are far from being there yet, but we need to strive to keep the system designed this way(where this is not too costly) and not introduce new places that removes backpressure.
But I'll leave it up to you @mystenmark  - I don't think its the end of the world if we merge this as is if you are not convienced.

**Comment by mwtian** [comment]
> I still think all parts of the system need to support back pressure. We are far from being there yet, but we need to strive to keep the system designed this way(where this is not too costly) and not introduce new places that removes backpressure.

Totally agree with the system should support back pressure. In general, I think back pressure should go "outside" to clients eventually, and not go "inside" into Sui. I think consensus is the most "inner" part where itself does not really have a way to deal with back pressure from other components.

**Comment by mystenmark** [comment]
closed in favor of #9378

**Pull request closed without merging**
